Support sample-to-data-element links

jorainer opened the pull request:
This adds support for links between samples and data elements within an `MsExperiment`:
- add `LinkedMsExperiment` that supports links between samples and data elements
- add `linkSampleData` function to establish the links
- add `[, j` to support subsetting of an experiment by samples.

Documentation is not complete yet, but I suggest to evaluate/check the code in *Linking experimental data to samples* section in the vignette to get a feeling of how that links work - and provide feedback/change requests based on that.

jorainer committed:
feat: add LinkedMsExperiment and required functions (issue #17)

- Add `LinkedMsExperiment`.
- Add `linkSampleData` and related helper functions.

jorainer committed:
feat: add subsetting functionality

jorainer committed:
docs: add section for sample linking to the vignette

lgatto commented on vignettes/MsExperiment.Rmd:
```diff
@@ -250,6 +250,158 @@ sp
+
+Thus, while we added a single *annotation* file to the data element
+`"annotations"` in `experimentFiles`, after subsetting we ended up with two
+identical files. This eventual duplication of *n:m* relationships between
```
eventual isn't corrent here. I suggest you simply remove it.

jorainer replied:
fixed that.

lgatto commented on vignettes/MsExperiment.Rmd:
````diff
@@ -250,6 +250,158 @@ sp
+                      fromIndex = c(1, 2), toIndex = c(1, 2))
+```
+
+With parameter `with` we define which element in the object we want to link
````
This first sentence ins't clear - it might be missing a word somewhere. It's the first sentence that explains the linking, so it should be very clear.

jorainer replied:
I completely rewrote the whole section.

lgatto commented on vignettes/MsExperiment.Rmd:
```diff
@@ -250,6 +250,158 @@ sp
+samples to elements does however not affect data consistency. A sample will
+always be linked to the correct value/element.
+
+
```
I think a figure that draws the links between all the items could be helpful.

jorainer replied:
I added now 3 figures.

lgatto commented on R/LinkedMsExperiment.R:
```diff
@@ -0,0 +1,366 @@
+#'
+#' @exportMethod linkSampleData
+#'
+#' @importFrom S4Vectors List
```
Could you add a `@seealso` in the two `MsExperiment`-class man pages. Or wouldn't it be better to document both classes together?

jorainer replied:
Note that I did not finish the documentation of the `LinkedMsExperiment` - I wanted to first see if you would agree with the general concept and workflow before spending time on the documentation...

jorainer replied:
Now everything is documented.

lgatto commented on R/LinkedMsExperiment.R:
```diff
@@ -0,0 +1,366 @@
+#'
+#' @section Implementation notes:
+#'
+#' Links from samples to any other element are stored as an `integer` `matrix`
```
This first sentence is wrong I think. Something might be missing between `matrix` and *as a list ...*

jorainer replied:
Yep, sorry that I did not mention that the documentation is unfinished. I will fix all this later.

jorainer replied:
Documentation has been completed.

lgatto commented on R/LinkedMsExperiment.R:
```diff
@@ -0,0 +1,366 @@
+#' name needs to be in the format `<slot name>.<element name>`, e.g.
+#' `"experimentFiles.mzML_files"` to link samples to an element called
+#' `"mzML_files"` in `@experimentFiles`.
+#'
```
Shouldn't `fromIndex` and `toIndex` be documented in here too?

jorainer replied:
Sure, I will fix.

jorainer replied:
documentation added.

lgatto commented on R/LinkedMsExperiment.R:
```diff
@@ -0,0 +1,366 @@
+                   toIndex = integer()) {
+              object <- as(object, "LinkedMsExperiment")
+              if (!length(with))
+                  return(object)
```
Here, you return a `LinkedMsExperiment` even though no links have been added. Wouldn't it be safer to return the original `object` in such case?

jorainer replied:
I'm not sure. I think it would be confusing if `linkSampleData` returns in some instances a `MsExperiment` and in others `LinkedMsExperiment`. Returning a `LinkedMsExperiment` (even without links) would be more consistent.

jorainer replied:
All functionality is now implemented for `MsExperiment`, thus this should be fixed too.

lgatto commented:
### Handling of `MsExperiment` sample subsetting error

```r
> mse
Object of class MsExperiment 
 Files: mzML_files, annotations 
 Spectra: MS1 (1862) 
 Experiment data: 2 sample(s)
> mse[, 1]
Error in mse[, 1] : object of type 'S4' is not subsettable
```

while

```r
> mse <- linkSampleData(mse, with = "experimentFiles.mzML_file",
+                       fromIndex = c(1, 2), toIndex = c(1, 2))
> mse[, 1]
Object of class LinkedMsExperiment 
 Files: mzML_files, annotations, mzML_file 
 Spectra: MS1 (1862) 
 Experiment data: 1 sample(s)
 Sample data links:
  - experimentFiles.mzML_file: 1 sample(s) to 1 element(s).
```

I think it would be useful to gracefully catch the error for a non-linked experiment.

### Other data slots

- I assume that the same subsetting will also need to handle the other main data slots, (i.e. `QFeatures` and later `Chromatogram`)
- More of a discussion point rather than a request at this stage, but could this subsetting also handle data in the `@otherData` slot if the user defined these links?

jorainer commented:
Note that no `[` is implemented for `MsExperiment`. I implemented all the linking and subsetting for the `LinkedMsExperiment` class. The main discussion point will be if we want this functionality be implemented directly for the `MsExperiment` or if we want to keep it for the separate object?

And yes, I did not show it, but `linkSampleData` works also on the other slots! It works by simply using e.g. `with = "otherData.<element name>"`.

jorainer commented:
As discussed I will:
- move the functionality to `MsExperiment` and drop the `LinkedMsExperiment` object.
- `linkSampleData`: rename parameters `fromIndex` and `toIndex` to `sampleIndex` and `withIndex` and add a new parameter `subsetBy` to support subsetting by rows (default) or dimension 2, 3 etc.
- add the documentation
- add drawings illustrating each linking step in the vignette

jorainer committed:
small fix in vignette

jorainer committed:
Merge branch 'main' into jomain

jorainer committed:
refactor: move code to MsExperiment and drop LinkedMsExperiment

jorainer committed:
docs: add documentation and add content to vignette

jorainer committed:
Merge branch 'main' into jomain

jorainer committed:
refactor: adapt to renaming qfeatures -> assay

jorainer commented:
I've now added the documentation, removed the `LinkedMsExperiment` and implemented the functionality in `MsExperiment`. Also, linking and subsetting supports now the `@assay` slot with e.g. `SummarizedExperiment` - subsetting by sample is then done on columns. An example was added to the vignette.

lgatto commented on R/MsExperiment.R:
```diff
@@ -228,3 +368,78 @@ setReplaceMethod("metadata", "MsExperiment",
+#' @rdname MsExperiment
+#'
+#' @export
+setMethod(
```
Does this need the be a method with a generic function? Why not simply a function?

jorainer replied:
you're right. I made it a method because I thought I might implement for `MsExperiment` and `LinkedMsExperiment`, but no need for that anymore. I changed that in the last commit.

lgatto commented on R/MsExperiment.R:
```diff
@@ -228,3 +368,78 @@ setReplaceMethod("metadata", "MsExperiment",
                  })
+
+#' @rdname MsExperiment
+setGeneric("linkSampleData", function(object, ...)
```
I am not convinced that we gain anything from making this a generic + method. A simple function should suffice.

jorainer replied:
Yep. I changed that. Originally I wanted to implemet this for `MsExperiment` and `LinkedMsExperiment` but now (with `LinkedMsExperiment` gone) it's better to have a function.

lgatto commented on R/MsExperiment.R:
```diff
@@ -22,17 +26,22 @@ setClassUnion("QFeaturesOrSummarizedExperimentOrNull", c("NULL", "QFeaturesOrSum
 #'
+#' - Description and information (covariates etc) of each sample from the
+#'   experiment. These are stored in the `sampleData` slot as a `DataFrame`,
+#'   each row being a sample with columns containing all relevant informatio on
```
Each row **documenting** a sample with columns containing all relevant information**n** ...

jorainer replied:
agree

jorainer committed:
refactor: make linkSampleData a simple function

lgatto committed:
docs: fix/improve class man page

lgatto committed:
docs: rename ccl section

jorainer commented:
Btw, I've disabled all the code chunks from the first section of the vignette - @lgatto , I would ask you to check them again - they did not run on my computer and I was not sure how to fix them

lgatto commented on R/MsExperiment.R:
```diff
@@ -22,36 +26,137 @@ setClassUnion("QFeaturesOrSummarizedExperimentOrNull", c("NULL", "QFeaturesOrSum
 #'
+#' @section Linking sample data to other experimental data:
+#'
+#' By default, an `MsExperiment` is just a loose collection of files and data
```
May to 'To start with, an `MsExperiment` is just a ...'

jorainer replied:
OK

lgatto commented on R/MsExperiment.R:
```diff
@@ -22,36 +26,137 @@ setClassUnion("QFeaturesOrSummarizedExperimentOrNull", c("NULL", "QFeaturesOrSum
+#' By default, an `MsExperiment` is just a loose collection of files and data
+#' related to an experiment, no explicit links or associactions are present
+#' between the samples and related data. Such links can however be created with
+#' the `linkSampleData` function. This function can establish links between
```
with the `linkSampleData()` function

jorainer replied:
You mean adding `()` there? OK, did that.

lgatto commented on R/MsExperiment.R:
```diff
@@ -22,36 +26,137 @@ setClassUnion("QFeaturesOrSummarizedExperimentOrNull", c("NULL", "QFeaturesOrSum
+#' individual (or all) samples within the object's `sampleData` to individual,
+#' or multiple, data elements or files, such as `Spectra` or raw data files.
+#'
+#' The presence of such links enables a (coherent) subsetting of an
```
~~(coherent)~~ consistent

jorainer replied:
OK

lgatto commented on R/MsExperiment.R:
```diff
@@ -22,36 +26,137 @@ setClassUnion("QFeaturesOrSummarizedExperimentOrNull", c("NULL", "QFeaturesOrSum
+#'
+#' @param i for `[`: not supported.
+#'
+#' @param j for `[`: an `integer`, `character` or `logical` being the index, the
```
maybe rephrase 

> an `integer`, `character` or `logical` being the index, the name (rowname of `sampleData`) or a `logical` of the samples to subset.

as

> an `integer`, `character` or `logical` referring to the indices or names (rowname of `sampleData`) of the samples to subset.

jorainer replied:
sure, thanks

lgatto commented on vignettes/MsExperiment.Rmd:
````diff
@@ -244,12 +244,250 @@ We can access the raw data as long as the mzML files that were used to
+a single sample measured in two different injections to the same LC-MS setup.
+
+```{r}
+mse <- MsExperiment()
````
May be call this one `mse2` or something else to avoid any confusion with the one above.

jorainer replied:
renamed it to `lcmse` (LC-MS experiment).

lgatto reviewed (changes requested):
Here's a first set of suggestions. I will also look a bit more closely at the code, but it would be great to merge rather soon than later.

lgatto committed:
docs: eval code chunks

lgatto committed:
docs: eval first chunk (missed it previously) and explicitly load Spectra

jorainer committed:
docs: address Laurent's comments and suggestions

lgatto commented:
Thanks, will merge.

Two points
- I would suggest to rename `lcmse` (`mse` is also LC-MS - LC-MSMS, actually) to `lmse`, for linked experiment. 
- I think I will do some work on the vignette to merge the two parts and use only one set of data (probably the one from `msdata`).
